src/githubAuth.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Created by francesco on 19/01/17.
+ */
+
+const url = require('url');
+const _ = require('lodash');
+let request = require('request');
+
+request = request.defaults({
+  json: true,
+  headers: {
+    'User-Agent': 'LK Documentation'
+  }
+});
+
+class GithubAuth {
+  constructor(auth, domain) {
+    this.clientID = auth.clientID;
+    this.clientSecret = auth.clientSecret;
+    this.redirectUrl = auth.redirectUrl;
+    this.teamId = auth.teamId;
+    this.urlPrefix = auth.urlPrefix;
+    this.domain = domain;
+  }
+
+  authMiddleware(req, res, next) {
+    if (req.session.user &&
+        req.session.user.domains &&
+        req.session.user.domains.indexOf(this.domain) !== -1) {
+      next();
+    } else if (req.query.code) {
+      if (req.session.state !== req.query.state) {
+        // something fishy, or more simply someone is not using the right URL
+        res.status(400).send("req.session.state !== req.query.state");
+      } else {
+        request.post({
```

Comment:
Split this code in smaller one-purpose methods (e.g. `isAuthenticated(req)`, `authenticate(req, cb)`, `getUsername(cb)` `checkMembership(username, teamId, cb)`, etc.)